File: kb_modules/kb_typescript/postgres/kb_construct/src/ConstructKB.ts
// ConstructKB.ts
// ConstructKB.ts
import type { Client } from 'pg';
import { KnowledgeBaseManager, ConnectionParams } from './KnowledgeBaseManager';

export class ConstructKB extends KnowledgeBaseManager {
  public path: Record<string, string[]> = {};
  public pathValues: Record<string, Record<string, boolean>> = {};
  public workingKb?: string;

  constructor(
    host: string,
    port: number,
    dbname: string,
    user: string,
    password: string,
    tableName = 'knowledge_base',
  ) {
    const connParams: ConnectionParams = { host, port, database: dbname, user, password };
    super(tableName, connParams);
  }

  /**
   * Connects and creates tables (must be called before use)
   */
  async init(): Promise<void> {
    await super.init();
  }

  /**
   * Expose the raw pg Client
   */
  getDbObjects(): Client {
    return this.client;
  }

  /**
   * Add a new knowledge base and initialize its path stack
   */
  async addKb(kbName: string, description = ''): Promise<void> {
    if (this.path[kbName]) {
      throw new Error(`Knowledge base ${kbName} already exists`);
    }
    this.path[kbName] = [kbName];
    this.pathValues[kbName] = {};
    await super.addKb(kbName, description);
  }

  /**
   * Select which knowledge base to work on
   */
  selectKb(kbName: string): void {
    if (!this.path[kbName]) {
      throw new Error(`Knowledge base ${kbName} does not exist`);
    }
    this.workingKb = kbName;
  }

  /**
   * Add a header node under the current KB, pushing [link, nodeName] on the stack
   */
  async addHeaderNode(
    link: string,
    nodeName: string,
    nodeProperties: Record<string, any>,
    nodeData: Record<string, any>,
    description = '',
  ): Promise<void> {
    if (typeof description !== 'string') {
      throw new TypeError('description must be a string');
    }
    if (typeof nodeProperties !== 'object' || Array.isArray(nodeProperties)) {
      throw new TypeError('nodeProperties must be a dictionary');
    }
    if (!this.workingKb) {
      throw new Error('No knowledge base selected');
    }

    if (description) {
      nodeProperties.description = description;
    }

    this.path[this.workingKb].push(link, nodeName);
    const nodePath = this.path[this.workingKb].join('.');

    if (this.pathValues[this.workingKb][nodePath]) {
      throw new Error(`Path ${nodePath} already exists in knowledge base`);
    }
    this.pathValues[this.workingKb][nodePath] = true;

    await super.addNode(
      this.workingKb,
      link,
      nodeName,
      nodeProperties,
      nodeData,
      nodePath,
    );
  }

  /**
   * Add an info node and immediately pop back up one level
   */
  async addInfoNode(
    link: string,
    nodeName: string,
    nodeProperties: Record<string, any>,
    nodeData: Record<string, any>,
    description = '',
  ): Promise<void> {
    await this.addHeaderNode(link, nodeName, nodeProperties, nodeData, description);
    if (this.workingKb) {
      this.path[this.workingKb].pop(); // nodeName
      this.path[this.workingKb].pop(); // link
    }
  }

  /**
   * Leave the current header node, verifying the popped values match
   */
  leaveHeaderNode(label: string, name: string): void {
    if (!this.workingKb) {
      throw new Error('No knowledge base selected');
    }
    const stack = this.path[this.workingKb];
    if (stack.length < 3) {
      throw new Error('Cannot leave a header node: path is empty or too short');
    }

    const refName  = stack.pop()!;
    const refLabel = stack.pop()!;
    const errors: string[] = [];

    if (refName !== name)  errors.push(`Expected name '${name}', got '${refName}'`);
    if (refLabel !== label) errors.push(`Expected label '${label}', got '${refLabel}'`);
    if (errors.length) {
      // restore state
      stack.push(refLabel, refName);
      throw new Error(errors.join('; '));
    }
  }

  /**
   * Add a link node under the current path
   */
  async addLinkNode(linkName: string): Promise<void> {
    if (!this.workingKb) {
      throw new Error('No knowledge base selected');
    }
    const currentPath = this.path[this.workingKb].join('.');
    await super.addLink(this.workingKb, currentPath, linkName);
  }

  /**
   * Add a link mount under the current path
   */
  async addLinkMount(
    linkMountName: string,
    description = '',
  ): Promise<{ knowledge_base: string; mount_path: string }> {
    if (!this.workingKb) {
      throw new Error('No knowledge base selected');
    }
    const currentPath = this.path[this.workingKb].join('.');
    return await super.addLinkMount(
      this.workingKb,
      currentPath,
      linkMountName,
      description,
    );
  }

  /**
   * Verify that every KB path stack is back to its root, then disconnect.
   */
  async checkInstallation(): Promise<void> {
    for (const kbName in this.path) {
      const stack = this.path[kbName];
      if (stack.length !== 1 || stack[0] !== kbName) {
        throw new Error(
          `Installation check failed for '${kbName}'. Path: [${stack.join(', ')}]`
        );
      }
    }
    
  }
}

File: kb_modules/kb_rustc/kb_create/kb_create_module/src/ConstructKb.rs
// src/construct_kb.rs

use std::collections::HashMap;
use serde_json::Value;
use postgres::Error;
use base_construct_kb::KnowledgeBaseManager;

/// Builds on KnowledgeBaseManager to maintain header/info stacks per KB.
pub struct ConstructKb {
    inner: KnowledgeBaseManager,
    /// path stack for each KB (e.g. ["kb1", "link1", "node1", …])
    pub path: HashMap<String, Vec<String>>,
    /// tracks which full paths have been used per KB
    pub path_values: HashMap<String, HashMap<String, bool>>,
    working_kb: Option<String>,
}

impl ConstructKb {
    /// Connects and (re)creates tables, just like Python __init__.
    pub fn new(
        host: &str,
        port: u16,
        dbname: &str,
        user: &str,
        password: &str,
        table_name: &str,
    ) -> Result<Self, Error> {
        let conn_str = format!(
            "host={} port={} dbname={} user={} password={}",
            host, port, dbname, user, password
        );
        let mgr = KnowledgeBaseManager::new(table_name, &conn_str)?;
        Ok(Self {
            inner: mgr,
            path: HashMap::new(),
            path_values: HashMap::new(),
            working_kb: None,
        })
    }

    /// Expose the raw client & transaction if you need it.
    pub fn get_db_objects(&mut self) -> (&mut postgres::Client, &mut postgres::Transaction) {
        self.inner.get_db_objects()
    }

    /// Mirror Python add_kb: initialize stacks and call base.
    pub fn add_kb(&mut self, kb_name: &str, description: &str) -> Result<(), Error> {
        if self.path.contains_key(kb_name) {
            return Err(Error::from(std::io::Error::new(
                std::io::ErrorKind::AlreadyExists,
                format!("KB '{}' already exists", kb_name),
            )));
        }
        self.inner.add_kb(kb_name, Some(description))?;
        self.path.insert(kb_name.to_string(), vec![kb_name.to_string()]);
        self.path_values.insert(kb_name.to_string(), HashMap::new());
        Ok(())
    }

    /// Select which KB subsequent calls apply to.
    pub fn select_kb(&mut self, kb_name: &str) -> Result<(), Error> {
        if !self.path.contains_key(kb_name) {
            return Err(Error::from(std::io::Error::new(
                std::io::ErrorKind::NotFound,
                format!("KB '{}' does not exist", kb_name),
            )));
        }
        self.working_kb = Some(kb_name.to_string());
        Ok(())
    }

    /// Push a header node, record it, and call base.add_node.
    pub fn add_header_node(
        &mut self,
        link: &str,
        node_name: &str,
        node_properties: &mut HashMap<String, Value>,
        node_data: &Value,
        description: Option<&str>,
    ) -> Result<(), Error> {
        let kb = self.working_kb.as_ref().ok_or_else(|| {
            Error::from(std::io::Error::new(std::io::ErrorKind::InvalidInput, "No KB selected"))
        })?;

        // optionally attach description
        if let Some(desc) = description {
            node_properties.insert("description".into(), Value::String(desc.into()));
        }

        let stack = self.path.get_mut(kb).unwrap();
        stack.push(link.into());
        stack.push(node_name.into());
        let full_path = stack.join(".");

        let vals = self.path_values.get_mut(kb).unwrap();
        if vals.contains_key(&full_path) {
            return Err(Error::from(std::io::Error::new(
                std::io::ErrorKind::AlreadyExists,
                format!("Path '{}' already exists", full_path),
            )));
        }
        vals.insert(full_path.clone(), true);

        self.inner.add_node(
            kb,
            link,
            node_name,
            Some(&Value::Object(node_properties.clone())),
            Some(node_data),
            &full_path,
        )?;
        Ok(())
    }

    /// Add header node then immediately pop it off.
    pub fn add_info_node(
        &mut self,
        link: &str,
        node_name: &str,
        node_properties: &mut HashMap<String, Value>,
        node_data: &Value,
        description: Option<&str>,
    ) -> Result<(), Error> {
        self.add_header_node(link, node_name, node_properties, node_data, description)?;
        let kb = self.working_kb.as_ref().unwrap();
        let stack = self.path.get_mut(kb).unwrap();
        stack.pop();
        stack.pop();
        Ok(())
    }

    /// Pop a header, verifying it matches the expected label & name.
    pub fn leave_header_node(&mut self, label: &str, name: &str) -> Result<(), Error> {
        let kb = self.working_kb.as_ref().ok_or_else(|| {
            Error::from(std::io::Error::new(std::io::ErrorKind::InvalidInput, "No KB selected"))
        })?;
        let stack = self.path.get_mut(kb).unwrap();
        if stack.len() < 3 {
            return Err(Error::from(std::io::Error::new(
                std::io::ErrorKind::Other,
                "Not enough elements to leave header",
            )));
        }
        let popped_name = stack.pop().unwrap();
        let popped_label = stack.pop().unwrap();
        if popped_name != name || popped_label != label {
            return Err(Error::from(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!(
                    "Expected label/name '{}/{}' but got '{}/{}'",
                    label, name, popped_label, popped_name
                ),
            )));
        }
        Ok(())
    }

    /// Create a link under the current header path.
    pub fn add_link_node(&mut self, link_name: &str) -> Result<(), Error> {
        let kb = self.working_kb.as_ref().unwrap();
        let full_path = self.path.get(kb).unwrap().join(".");
        self.inner.add_link(kb, &full_path, link_name)
    }

    /// Create a link‐mount under the current header path.
    pub fn add_link_mount(
        &mut self,
        link_mount_name: &str,
        description: Option<&str>,
    ) -> Result<(), Error> {
        let kb = self.working_kb.as_ref().unwrap();
        let full_path = self.path.get(kb).unwrap().join(".");
        self.inner
            .add_link_mount(kb, &full_path, link_mount_name, description.unwrap_or(""))
            .map(|_| ())
    }

    /// Verify all stacks are reset to just the KB name.
    pub fn check_installation(&self) -> Result<bool, Error> {
        for (kb, stack) in &self.path {
            if stack.len() != 1 || stack[0] != *kb {
                return Err(Error::from(std::io::Error::new(
                    std::io::ErrorKind::Other,
                    format!("Installation check failed for KB {:?}: {:?}", kb, stack),
                )));
            }
        }
        Ok(true)
    }

    /// Cleanly close the connection.
    pub fn disconnect(self) {
        self.inner.disconnect();
    }
}
